doc/configuration.md
# Configuration

Program configuration is done with a configuration file. The file
location must be specified on the command line. To ease specifying new
configurations, a [template file](../template.toml) is provided. The
syntax is [toml](https://github.com/toml-lang/toml), an ini-like
language for use in configuration files.

There are several configuration sections in the file:

* [Source](#section-source): where and how to retrieve source files
* [CSP](#section-csp): how to handle CSP files (if any)
* [Data](#section-data): how to handle data files (if any)
* [GitHub](#section-github): how to handle sources on GitHub
* [Bitbucket](#section-bitbucket): how to handle sources on Bitbucket
* [Directory](#section-directory): how to handle sources on the
  filesystem
* [Server](#section-server): details about the UDL to XML conversion
  server
* [Local](#section-local): details about the output file, logging, etc.

## Section Source

* **type**: the location of the sources; can be `'github'` (configured
  under [GitHub](#section-github)), `'bitbucket'` (configured under
  [Bitbucket](#section-bitbucket)), or `'directory'` (configured under
  [Directory](#section-directory)).

* **encoding**: the encoding of the sources. Should normally be `'UTF-8'`
  (which is the default). Can be any valid [Python encoding
  specifier](https://docs.python.org/3.9/library/codecs.html#standard-encodings).

* **srctype**: whether the sources are in `'udl'` (the default) or
  `'xml'`.

* **srcdir**: a path relative to the sources root where source files can
  be found, e.g. `'src'`.

* **cspdir**: if CSP files are present, they need to be placed in a
  separate directory, as they are handled differently than normal source
  files. A path relative to the sources root, e.g. `'csp'`. If empty, no
  handling of CSP files is done.

* **datadir**: if data files (e.g., lookup tables) are present, they
  need to be placed in a separate directory, as they are handled
  differently than normal source files. A path relative to the sources
  root, e.g. `'data'`.

* **skip**: a list (enclosed in square brackets) of items to exclude
  from the export. Items are specified as a path (including leading
  slash) from the sources root. Supports asterisk-style wildcards.
  Example: `['/src/tests/*','/data/Test_*.lut']`

* **take**: a list (enclosed in square brackets) of items to include in
  the export. Items are specified as a path (including leading slash)
  from the sources root. Supports asterisk-style wildcards. Example:
  `['/src/tests/*','/data/Test_*.lut']`. If this list is empty, all
  items will be included. If an item is present in both the `skip` and
  `take` lists (or the latter is empty), the `skip` list takes
  precedence.

## Section CSP

(This section and the CSP.parsers sections are only needed if CSP files
are present, i.e., the `cspdir` setting under [Source](#section-source)
is nonempty.)

Including CSP files in an XML export is somewhat challenging. In the
sources root, a CSP file is just a path and filename. However, in an XML
export, the path to the file is split into two parts: an _application_
and an _item name_. The first part describes the CSP application the
file belongs to, and the second part the path and filename _within that
application_.

The challenge, therefore, is to split a full path and filename from the
sources root, into an application and item. This is what the
[CSP.parsers](#section-cspparsers) subsections are intended for. There can be
one or more of these sections, which is denoted by them being enclosed
in double square brackets (i.e., `[[CSP.parsers]]`).

This general section has only one, general, option:

* **export**: how to export CSP items. There are three possible values:
  * `'embed'` (the default) places the items in the same export file as
    the source items.
  * `'separate'` creates a separate export file for the CSP items. It will
    have the same name as the source items export, with `_csp` appended.
  * `'none'` skips all CSP items.

### Section CSP.parsers

CSP parsers map the path of a CSP item in the source tree to an
application and item within that application. More than one of these
sections can be present; this is denoted by the double brackets that
surround the section in the configuration.

The parsers use a [regular
expression](https://en.wikipedia.org/wiki/Regular_expression) (specified
in **regex**) to determine which part of a CSP path is the application,
and which part the actual item. Note that replacements can be hardcoded
if needed. Some [examples](#examples) can be found below.

The settings for each parser are:

* **regex**: a regular expression to match against the item path.

* **app**: what to use as CSP application when the regex matches. Can be
  a back reference or a literal string. The application should start
  with a slash, and may not end with it.

* **item**: what to use as the item name within the application. Can be
  a back reference or a literal string. The item may not start with a
  slash.

* **nomatch**: what to do if this parser does not match the item. One of
  `'error'` to stop processing with an error message, or `'skip'` to
  continue with the next parser (if any). If the last parser doesn't
  match and specifies `'skip'`, the item in question will not be
  included in the export.

#### Examples

The default parser specified in the template configuration is this:

```toml
[[CSP.parsers]]
regex = '((/csp)?/[^/]+)/(.+)'
app = '\1'
item = '\3'
nomatch = 'error'
```

This splits:

* `/app/page.csp` in `/app` and `page.csp`
* `/app/sub/dir/page.csp` in `/app` and `sub/dir/page.csp`
* `/csp/app/dir/page.csp` in `/csp/app` and `dir/page.csp`

That is, it assumes the first part is the application, unless that part
is `csp`, in which case it uses the first two parts as application. In
both cases, anything left is used as the item name.

The parser above tries to be as universal as possible. In most cases,
there is only a single CSP application in a given configuration, with a
known path. In this case, assuming that CSP application is called
`/app`, the following configuration could be used:

```toml
[[CSP.parsers]]
regex = '/app/(.+)'
app = '/app'
item = '\1'
nomatch = 'error'
```

## Section Data

This section describes how data exports are handled. The is one setting:

* **export**: how to export data items. There are three possible values:
  * `'embed'` (the default) places the items in the same export file as
    the source items.
  * `'separate'` creates a separate export file for the data items. It
    will have the same name as the source items export, with `_data`
    appended.
  * `'none'` skips all data items.

## Section GitHub

This section describes where to find sources located on GitHub. They
will be downloaded automatically.

* **user**: the GitHub username, i.e. the repository owner

* **repo**: the GitHub repository name, i.e. the actual repository

* **tag**: the tag to get the sources for. This may be anything that
  GitHub supports, i.e. a tag, branch name, commit id, etc.

* **token**: if the GitHub repository is private, and a token is needed
  to access it, it can be placed here. Information about creating a
  GitHub access token can be found
  [here](https://docs.github.com/en/github/authenticating-to-github/keeping-your-account-and-data-secure/creating-a-personal-access-token).
  The token needs only `repo` permissions.
  
  If the value starts with an at-sign, it is interpreted as a path to a
  file containing the token. This can be useful to share the
  configuration file with others, without accidentally sharing a
  personal GitHub token. The path may be specified absolute, or relative
  to the configuration file. Note that setting `augment_from` in section
  `[Local]` can also be used to specify an external token.

## Section Bitbucket

This section describes where to find sources located on Bitbucket. They
will be downloaded automatically.

* **owner**: the Bitbucket repository owner

* **repo**: the Bitbucket repository name

* **tag**: the tag to get the sources for. This may be anything that
  Bitbucket supports, i.e. a tag, branch name, commit id, etc.

* **user** and **token**: if the Bitbucket repository is private,
  specify the username and [app
  password](https://support.atlassian.com/bitbucket-cloud/docs/create-an-app-password/)
  here. The app password only needs `Repositories:Read` permissions.

## Section Directory

If sources are to be found in a filesystem directory, this section
specifies where exactly to find them. It has only one option:

* **path**: the path to the source files. If not an absolute path, it
  will be interpreted as relative to the configuration file.

## Section Server

The Server section is only needed if the sources are stored in UDL.
Exports are always XML, so these sources need to be converted. This is
done by calling an API on a Cache/IRIS server, in web application
`/api/atelier`. The credentials supplied (see below) should therefore be
sufficient to access that application.

Configuration options are:

* **host**: the hostname of the server; default is `'localhost'`.

* **port**: the webserver port of the server; default is `'52773'`.

* **namespace**: the namespace to perform the conversion in. Only
  matters if some namespaces are locked down; default is `'USER'`.

* **user**: username to use accessing the API; default is `'SuperUser'`.

* **password**: password to use accessing the API; default is `'SYS'`.

* **https**: whether to use HTTPS to access the server; boolean, default
  is `false`.

## Section Local

This section specifies where to create the export file, what and
where to log, and configures various miscellaneous settings.

* **outfile**: the name of the output file to create. This is a string
  supporting a few replacements:

  * `{name}`: will be replaced with the top-level directory in the
    filesystem or GitHub/Bitbucket zip.
  * `{cfgname}`: will be replaced with the name (without extension) of the
    configuration file.
  * `{tag}`: will be replaced with the tag as configured in the GitHub
    or Bitbucket section. This is mostly useful with Bitbucket, where
    `{name}` will always be a commit id.
  * `{timestamp}`: will be replaced with the current date/time in format
    `YYYY-MM-DDThh-mm`.

* **deployment**: a boolean (default `false`) indicating whether the
  export should be created as an Ensemble deployment. This adds a Studio
  project, and a `ProjectTextDocument` XML element that Ensemble uses
  when processing a deployment. Importing a deployment has the advantage
  that an import log is maintained, and rollback functionality provided.
  Support for deployments here is minimal, though, as there is no
  documentation. Additionally, deployments do not support CSP items.

* **logdir**: the directory to place the log file in. Can be specified
  absolute or relative to the configuration file. The logfile itself has
  the same name as the configuration file, with a `.log` extension. If
  no directory is specified, the log file is placed adjacent to the
  configuration file.

* **loglevel**: the level of logging. Possible values are `'debug'`,
  `'info'`, `'warning'`, `'error'`, and `'critical'`. Note that most
  logging is done as `'info'` or `'debug'`.

* **timestamps**: how to handle timestamps (TimeCreated and TimeChanged
  elements) in classes. XML sources may or may not have them. Sources on
  the filesystem can use the file timestamp to set these properties.
  Possible values are:

  * `'clear'`: remove timestamp elements from class exports, if present.
  * `'update'`: if timestamp information can be determined, as on a
    local filesystem, add or update the elements, otherwise remove them.
  * `'leave'`: don't add or remove timestamps, leave them as they are.
    This is the default.
  
  UDL sources from GitHub/Bitbucket don't have timestamps, but they get
  one from the UDL to XML conversion. These are always stripped.

* **export_version**: specifies the version to use in the generated
  export file. Possible values are `25` (Cache) and `26` (IRIS). This is
  especially useful for UDL sources. If not specified here, the version
  used will be the one the UDL to XML conversion server creates. If the
  sources are actually saved from a Caché server, and the conversion
  server is IRIS (or vice versa), this can lead to problems with class
  storage definitions.
  
  For XML sources, this setting is usually best left empty.

* **sort**: a boolean (default `false`) indicating whether the items in
  the export should be sorted alphabetically. This can be useful for
  comparing exports from different sources, e.g. one created from a
  GitHub commit, with another one from a local (filesystem) repository.

* **converter**: how to convert UDL sources to XML. Can be `iris` (the
  default), which uses the IRIS server specified in the `[Server]`
  section, or `builtin`, which uses a built-in converter.

* **augment_from** (filename) specifies an external toml file to merge
  into the current one. It can be used to specify settings used in
  multiple toml files once. The toml file must have the same structure
  as the regular one, but does not have to be complete.


doc/releasenotes.md
# Release notes

## Version 0.8

Version 0.8 switches to a different way of distributing the program.
Previous versions used a
[PyInstaller](https://pyinstaller.org/en/stable/index.html) [single file
executable](https://pyinstaller.org/en/stable/operating-mode.html#bundling-to-one-file).
This mode of operation is slightly slower than the alternative, with the
program and supporting files placed in a directory. Version 0.8 starts
using the latter way of distributing.

Further changes in this release:

- Add an option to use a built-in UDL to XML converter, using
  [iris-udl-to-xml](https://github.com/gertjanklein/iris-udl-to-xml).
  This can be especially useful in CI/CD pipelines, as an IRIS server is
  no longer required.
- Add a [sort](../doc/configuration.md#section-local) setting, allowing
  to sort items in the export file.
- Bugfix: actually use https if so configured.
- The embedded Python is upgraded to version 3.12.

## Version 0.7.1

This version adds support for loading sources from Bitbucket.

## Version 0.6

This version adds:

- The [augment_from](../doc/configuration.md#section-local) setting,
  allowing common system-specific settings to be maintained in a shared
  configuration file. This option replaces the `take_from` setting.
- The [export_version](../doc/configuration.md#section-local) (expert)
  setting, allowing to override the version number in the export file.


src/config.py

import sys
import os
from os.path import exists, isfile, isabs, dirname, abspath, join, splitext, basename
import re
import datetime
import argparse
from io import StringIO
import logging

from typing import cast

import toml

import namespace as ns
from namespace import ConfigurationError


def get_config() -> ns.Namespace:
    """Parse the config file, check validity, return as Namespace."""
    
    # Get configuration filename from commandline
    args = parse_args()
    cfgfile = args.config

    # Initial logging setup: file next to config file. Errors parsing the
    # config file will be logged here.
    setup_basic_logging(cfgfile)
    
    # Log unhandled exceptions
    sys.excepthook = unhandled_exception

    # Load configuration from file
    config = ns.dict2ns(toml.load(cfgfile))

    # Add config file and directory
    config.cfgfile = cfgfile
    config.cfgdir = dirname(cfgfile)

    # Minimal check for logging configuration
    local = ns.check_section(config, 'Local')
    ns.check_default(local, 'logdir', '')
    levels = 'debug,info,warning,error,critical'.split(',')
    ns.check_oneof(local, 'loglevel', levels, 'info')

    # Merge-in setting from the file specified in augment_from, if any
    merge_augmented_settings(config)

    # Do final setup of logging
    setup_logging(config)

    # Merge command line overrides into configuration
    merge_overrides(args, config)

    # Make sure configuration is complete
    check(config)
    
    # # Converts specification(s) of files to skip/take to regexes
    for src, dst in (('skip', 'skip_regexes'), ('take', 'take_regexes')):
        config[dst] = []
        for spec in config.Source[src]:
            spec = spec.replace('\\', '\\\\')
            spec = spec.replace('.', '\\.')
            # Create valid regex for star
            spec = spec.replace('*', '.*')
            regex = re.compile(spec, re.I)
            config[dst].append(regex)
    
    # Load token contents from file, if specified
    if config.Source.type == 'github' and config.GitHub.token and config.GitHub.token[0] == '@':
        path = config.GitHub.token[1:]
        if not isabs(path):
            path = join(abspath(config.cfgdir), path)
        try:
            with open(path, encoding='UTF-8') as f:
                config.GitHub.token = f.read().strip()
        except OSError as e:
            raise ConfigurationError(f"Error reading token file {path}: {e}") from None
    
    return config


# =====

def merge_augmented_settings(config:ns.Namespace):
    """ Merges settings from file in setting augment_from, if any """
    
    local = ns.get_section(config, 'Local')
    if local is None:
        return
    fname = local._get('augment_from')
    if fname is None:
        return
    if not isabs(fname):
        fname = join(config.cfgdir, fname)
    if not exists(fname):
        raise ConfigurationError(f"augment_from file {local._get('augment_from')} not found")
    cs = ns.dict2ns(toml.load(fname))
    if (aug_local := ns.get_section(cs, 'Local')) and 'augment_from' in aug_local:
        raise ConfigurationError("Recursive augment_from not supported")
    # Add/override each key/value in augment_from
    for k, v in cs._flattened():
        ns.set_in_path(config, k, v)


def merge_overrides(args:argparse.Namespace, config:ns.Namespace):
    """Merge command line overrides into configuration"""
    
    config.no_gui = args.no_gui
    for arg in ARGS:
        value = getattr(args, cast(str, arg['name']))
        if not value:
            continue
        ns.set_in_path(config, cast(str, arg['path']), value)
        

def check(config:ns.Namespace):
    """Check validity of values in the parsed configuration."""

    # We need at least the Source and Config sections
    for name in ('Source', 'Local'):
        ns.check_section(config, name)
    
    # Check Source
    src = config.Source
    ns.check_oneof(src, 'type', ('github', 'bitbucket', 'directory'))
    ns.check_oneof(src, 'srctype', ('xml', 'udl'), 'udl')
    ns.check_encoding(src, 'encoding', 'UTF-8')
    
    # Set some defaults if needed
    ns.check_default(src, 'srcdir', '')
    ns.check_default(src, 'datadir', '')
    ns.check_default(src, 'cspdir', '')
    ns.check_default(src, 'skip', [])
    ns.check_default(src, 'take', [])
    
    # Strip leading slash if present, we don't need it
    if src.srcdir == '/':
        src.srcdir = src.srcdir[1:]
    if src.datadir == '/':
        src.datadir = src.datadir[1:]
    if src.cspdir == '/':
        src.cspdir = src.cspdir[1:]

    # Check Local section
    local = config.Local
    ns.check_notempty(local, 'outfile')
    ns.check_default(local, 'deployment', False)
    ns.check_default(local, 'logdir', '')
    ns.check_default(local, 'loglevel', '')
    ns.check_default(local, 'threads', 1)
    ns.check_default(local, 'sort', False)
    ns.check_oneof(local, 'timestamps', ('clear', 'update', 'leave'), 'leave')
    ns.check_oneof(local, 'export_version', (25, 26, '', None), '')
    if local.export_version == '':
        local.export_version = None
    ns.check_oneof(local, 'converter', ('iris', 'builtin'), 'iris')
    
    # Check data configuration
    if src.datadir:
        data = ns.get_section(config, 'Data', True)
        assert data is not None # stop mypy complaints
        ns.check_oneof(data, 'export', ('embed', 'separate', 'none'), 'embed')

    # Check CSP configuration
    if src.cspdir:
        check_csp(config)
    
        # CSP items appear unsupported in deployments, so must be exported separately
        if local.deployment and config.CSP.export == 'embed':
            raise ConfigurationError("When requesting a deployment, CSP export must be 'separate'.")
    
    # Check optional sections
    if src.type == 'directory':
        ns.check_section(config, 'Directory')
        ns.check_notempty(config.Directory, 'path')
        if 'structure' in config.Directory:
            logging.warning("Warning: setting 'structure' in section Directory no longer used.")
            del config.Directory.structure
        if not isabs(config.Directory.path):
            config.Directory.path = join(abspath(config.cfgdir), config.Directory.path)
    elif src.type == 'bitbucket':
        ns.check_section(config, 'Bitbucket')
        bb = config.Bitbucket
        ns.check_notempty(bb, 'owner')
        ns.check_notempty(bb, 'repo')
        ns.check_notempty(bb, 'tag')
        ns.check_default(config.Bitbucket, 'token', '')
        ns.check_default(config.Bitbucket, 'user', '')
    else:
        ns.check_section(config, 'GitHub')
        gh = config.GitHub
        ns.check_notempty(gh, 'user')
        ns.check_notempty(gh, 'repo')
        ns.check_notempty(gh, 'tag')
        ns.check_default(config.GitHub, 'token', '')

    # Check server configuration
    if src.srctype == 'udl':
        check_server(config)


def check_server(config):
    """ Check UDL -> XML server configuration """
    
    # Make sure a section is present; it may not be, as every
    # setting here has a default.
    svr = ns.get_section(config, 'Server')
    if svr is None:
        svr = config.Server = ns.Namespace()
    
    # Make sure default values are present
    ns.check_default(svr, 'host', 'localhost')
    ns.check_default(svr, 'port', '52773')
    ns.check_default(svr, 'user', 'SuperUser')
    ns.check_default(svr, 'password', 'SYS')
    ns.check_default(svr, 'namespace', 'USER')
    ns.check_default(svr, 'https', False)
    
    if 'take_from' in svr:
        raise ConfigurationError("Setting take_from no longer supported." \
            " Use augment_from in section Local instead.")


def check_csp(config:ns.Namespace):
    """ Checks the CSP configuration """

    csp = ns.check_section(config, 'CSP')
    ns.check_oneof(csp, 'export', ('embed', 'separate', 'none'), 'embed')
    if not csp.export == 'none':
        # Only check these if we are to export CSP files
        if ns.check_default(csp, 'parsers', []):
            msg = "At least one [[CSP.parsers]] section for CSP items should be present."
            raise ConfigurationError(msg)
        for i, parser in enumerate(csp.parsers):
            if not isinstance(parser, ns.Namespace):
                raise ConfigurationError(f'Parser {i+1} must be a section.')
            ns.check_notempty(parser, 'regex')
            ns.check_notempty(parser, 'app')
            ns.check_notempty(parser, 'item')
            ns.check_oneof(parser, 'nomatch', ('skip', 'error'), 'error')
    
# ==========

def setup_basic_logging(cfgfile):
    """ Initial logging setup: log to file next to config file """

    # Determine log file name
    base, ext = splitext(cfgfile)
    if ext.lower() == '.toml':
        logfile = f'{base}.log'
    else:
        logfile = f'{cfgfile}.log'
    
    # Create handler with delayed creation of log file
    handlers = [logging.FileHandler(logfile, delay=True)]

    # Display what we log as-is, no level strings etc.
    logging.basicConfig(handlers=handlers, level=logging.INFO,
        format='%(message)s', force=True)


def setup_logging(config):
    """ Final logging setup: allow log location override in config """

    logdir:str = config.Local.logdir
    loglevel:str = config.Local.loglevel

    # Determine filename (without path)
    base, ext = splitext(basename(config.cfgfile))
    if ext.lower() == '.toml':
        logfile = f'{base}.log'
    else:
        logfile = f'{base}{ext}.log'

    # Determine filename (with path)
    name = join(logdir, logfile)
    if not isabs(logdir):
        # Logdir not absolute: make it relative to dir config file is in
        name = join(dirname(config.cfgfile), name)

    # Make sure the log directory exists
    logdir = dirname(name)
    os.makedirs(logdir, exist_ok=True)

    # Replace the current logging handler with one using the newly
    # determined path
    logger = logging.getLogger()
    logger.handlers.clear()
    logger.handlers.append(logging.FileHandler(name, 'a'))

    if loglevel is not None:
        logger.setLevel(loglevel.upper())

    # Log appends; create visible separation for this run
    now = str(datetime.datetime.now()).split('.')[0] # pylint:disable=use-maxsplit-arg
    logging.info("\n\n===== Starting at %s", now)
    
# =====

def unhandled_exception(exc_type, exc_value, exc_traceback):
    """ Handle otherwise unhandled exceptions by logging them """

    if exc_type == ConfigurationError:
        msg = exc_value.args[0]
        logging.error("\n%s", msg)
    else:
        msg = f"An error occurred; please see the log file for details.\n\n{exc_value}"
        exc_info = exc_type, exc_value, exc_traceback
        logging.exception("\n##### Unhandled exception:", exc_info=exc_info)
    msgbox(msg, True)
    sys.exit(1)

# =====

def msgbox(msg, is_error=False):
    """ Display, if on Windows, a message box """

    if os.name == 'nt':
        if is_error:
            flags = 0x30
            title = "Error"
        else:
            flags = 0
            title = "Info"
        import ctypes
        MessageBox = ctypes.windll.user32.MessageBoxW
        MessageBox(None, msg, title, flags)
    else:
        print(msg)

# =====

# Command line overrides for values in the configuration file
ARGS = [
    {
        'name': 'github_tag',
        'path': 'GitHub.tag',
        'argparse': (
            ["--github-tag"],
            {
                'default': '',
                'help': "Override tag/branch to retrieve on GitHub"
            }
        )
    }
]

def parse_args():
    """Parse command line arguments."""

    parser = argparse.ArgumentParser()
    parser.add_argument("config",
       help="The (TOML) configuration file to use")
    parser.add_argument("--no-gui", action='store_true',
       help="Do not display a message box on completion.")
    
    # Add command line overrides
    for arg in ARGS:
        names, kwargs = arg['argparse']
        parser.add_argument(*names, **kwargs)

    # Replace stdout/stderr to capture argparse output
    sys.stdout = StringIO()
    sys.stderr = StringIO()
    
    # Check command line
    try:
        args = parser.parse_args()

    except SystemExit:
        # Get argparse output; either an error message in stderr, or
        # a usage message in stdout.
        msg, err = sys.stderr.getvalue(), True
        if not msg:
            msg = sys.stdout.getvalue()
            err = False
        
        # Show error or usage and exit
        msgbox(msg, err)
        raise

    finally:
        # Restore stdout/stderr
        sys.stdout = sys.__stdout__
        sys.stderr = sys.__stderr__
    
    cfgfile = args.config
    if not isabs(cfgfile) and not exists(cfgfile):
        cfgfile = join(dirname(__file__), cfgfile)
    
    if not exists(cfgfile) or not isfile(cfgfile):
        msgbox(f"Error: file {args.config} not found.\n\n{parser.format_help()}", True)
        sys.exit(1)
    
    if not isabs(cfgfile):
        cfgfile = abspath(cfgfile)
    
    args.config = cfgfile

    return args
